src/multiple_shooting.jl

```diff
@@ -126,6 +146,118 @@ function multiple_shoot(
+        throw(DomainError(group_size, "group_size can't be < 2 or > number of data points"))
+    end
+
+    # Get ranges that partition data to groups of size group_size
+    ranges = group_ranges(datasize, group_size)
+
+    # Multiple shooting predictions
+    # by using map we avoid mutating an array
+    sols =  map(
+        rg -> begin
+            newprob = remake(
+                    prob;
+                    p=p,
+                    tspan=(tsteps[first(rg)], tsteps[last(rg)]),
+                    );
+            function prob_func(prob, i, repeat)
+                remake(prob, u0 = ode_data[:, first(rg), i])
+            end
+            newensembleprob = EnsembleProblem(newprob, 
+                                            prob_func, 
+                                            ensembleprob.output_func,
+                                            ensembleprob.reduction,
+                                            ensembleprob.u_init,
+                                            ensembleprob.safetycopy);
+            solve(newensembleprob,
+                solver,
+                ensemblealg;
+                saveat=tsteps[rg],
+                kwargs...
+            )
+        end,
+        ranges)
+    group_predictions = Array.(sols)
+
+    # Abort and return infinite loss if one of the integrations did not converge?
+    convergeds = [sol.converged for sol in sols]
+    if any(.! convergeds)
+        return Inf, group_predictions
+    end
+
+    # Calculate multiple shooting loss
+    loss = 0
+    for (i, rg) in enumerate(ranges)
+        û = group_predictions[i]
+        u = ode_data[:, rg, :] # trajectories are at dims 3
+        # just summing up losses for all trajectories
+        # but other alternatives might be considered
+
+        loss += loss_function(u, û)
+
+        if i > 1
+            # Ensure continuity between last state in previous prediction
+            # and current initial condition in ode_data
+            loss +=
+                continuity_term * continuity_loss(group_predictions[i - 1][:, end, :], u[:, 1, :])
+        end
+    end
+
+    return loss, group_predictions
+end
+
+function multiple_shoot(
+    p::AbstractArray,
+    ode_data::AbstractArray,
+    tsteps::AbstractArray,
+    ensembleprob::EnsembleProblem,
+    ensemblealg::SciMLBase.BasicEnsembleAlgorithm,
+    loss_function::Function,
+    solver::DiffEqBase.AbstractODEAlgorithm,
+    group_size::Integer;
+    continuity_term::Real=100,
+    kwargs...
+)
+    # Continuity loss between last state in previous prediction
+    # and current initial condition in ode_data
+    function continuity_loss(û_end, u_0)
+        return sum(abs, û_end - u_0)
+    end
+
+    return multiple_shoot(
+        p,
+        ode_data,
+        tsteps,
+        ensembleprob,
+        ensemblealg,
+        loss_function,
+        continuity_loss,
+        solver,
+        group_size;
+        continuity_term,
+        kwargs...
+    )
+    
+end
+
```

Comment:
we should really move the optional argument to last

Reply:
I am not sure what you mean by that.

Reply:
the reason for this dispatch is to handle the optional argument, but it really shouldn't be one of the ones in the middle then.